File: backend/controllers/transactionController.js
const Transaction = require('../models/Transaction');
const Category = require('../models/Category');

// @desc    Create a new transaction
// @route   POST /api/transactions
// @access  Private
exports.createTransaction = async (req, res) => {
  try {
    const { amount, categoryId, date, note, type } = req.body;
    const userId = req.user.userId;

    const transaction = new Transaction({
      userId,
      amount,
      categoryId,
      date,
      note,
      type,
    });

    await transaction.save();
    res.status(201).json(transaction);
  } catch (error) {
    console.error('Error creating transaction:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Get all transactions for the user (with pagination & filtering)
// @route   GET /api/transactions
// @access  Private
exports.getTransactions = async (req, res) => {
  try {
    const userId = req.user.userId;
    const { page = 1, limit = 10, startDate, endDate, month, year } = req.query;

    const query = { userId };

    // Date filtering logic
    if (startDate && endDate) {
      query.date = { $gte: new Date(startDate), $lte: new Date(endDate) };
    } else if (month && year) {
      // Month is 1-indexed from the client (1 = January)
      const startOfMonth = new Date(year, month - 1, 1);
      const endOfMonth = new Date(year, month, 0, 23, 59, 59, 999);
      query.date = { $gte: startOfMonth, $lte: endOfMonth };
    }

    const skip = (parseInt(page) - 1) * parseInt(limit);

    const transactions = await Transaction.find(query)
      .sort({ date: -1 })
      .skip(skip)
      .limit(parseInt(limit))
      .populate('categoryId', 'name type icon color');

    const total = await Transaction.countDocuments(query);

    res.json({
      transactions,
      totalPages: Math.ceil(total / parseInt(limit)),
      currentPage: parseInt(page),
      total,
    });
  } catch (error) {
    console.error('Error fetching transactions:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Update a transaction
// @route   PUT /api/transactions/:id
// @access  Private
exports.updateTransaction = async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.userId;

    // Check if the transaction exists and belongs to the user
    let transaction = await Transaction.findOne({ _id: id, userId });
    if (!transaction) {
      return res.status(404).json({ message: 'Transaction not found or unauthorized' });
    }

    transaction = await Transaction.findByIdAndUpdate(
      id,
      { $set: req.body },
      { new: true, runValidators: true }
    );

    res.json(transaction);
  } catch (error) {
    console.error('Error updating transaction:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Delete a transaction
// @route   DELETE /api/transactions/:id
// @access  Private
exports.deleteTransaction = async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.userId;

    const transaction = await Transaction.findOneAndDelete({ _id: id, userId });
    if (!transaction) {
      return res.status(404).json({ message: 'Transaction not found or unauthorized' });
    }

    res.json({ message: 'Transaction deleted successfully' });
  } catch (error) {
    console.error('Error deleting transaction:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

File: backend/src/main/java/com/personalfinance/backend/controllers/TransactionController.java
package com.personalfinance.backend.controllers;

import com.personalfinance.backend.models.Transaction;
import com.personalfinance.backend.models.Category;
import com.personalfinance.backend.repositories.TransactionRepository;
import com.personalfinance.backend.repositories.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/transactions")
public class TransactionController {

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    @PostMapping
    public ResponseEntity<?> createTransaction(@RequestBody Transaction request) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        request.setUserId(userId);
        request.setDate(request.getDate() == null ? new Date() : request.getDate());
        Transaction saved = transactionRepository.save(request);
        return ResponseEntity.status(201).body(saved);
    }

    @GetMapping
    public ResponseEntity<?> getTransactions(@RequestParam(defaultValue = "1") int page,
                                             @RequestParam(defaultValue = "10") int limit,
                                             @RequestParam(required = false) Integer month,
                                             @RequestParam(required = false) Integer year) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();

        List<Transaction> transactions = transactionRepository.findByUserIdOrderByDateDesc(userId);

        if (month != null && year != null) {
            transactions = transactions.stream()
                .filter(t -> {
                    java.util.Calendar cal = java.util.Calendar.getInstance();
                    cal.setTime(t.getDate());
                    return cal.get(java.util.Calendar.MONTH) + 1 == month &&
                           cal.get(java.util.Calendar.YEAR) == year;
                })
                .collect(Collectors.toList());
        }

        int total = transactions.size();
        int startIndex = (page - 1) * limit;
        int endIndex = Math.min(startIndex + limit, total);
        if (startIndex > total) startIndex = total;
        List<Transaction> paginated = transactions.subList(startIndex, endIndex);

        Set<String> catIds = paginated.stream().map(Transaction::getCategoryId).collect(Collectors.toSet());
        System.out.println("Requested Category IDs: " + catIds);
        
        List<Category> cats = categoryRepository.findAllById(catIds);
        System.out.println("Found Categories: " + cats.size());
        
        Map<String, Category> catMap = new HashMap<>();
        for (Category c : cats) {
            catMap.put(c.getId(), c);
        }
        
        List<Map<String, Object>> resultList = new ArrayList<>();
        for (Transaction t : paginated) {
            Map<String, Object> map = new HashMap<>();
            map.put("_id", t.getId());
            map.put("userId", t.getUserId());
            map.put("amount", t.getAmount());
            map.put("date", t.getDate());
            map.put("note", t.getNote());
            map.put("type", t.getType());
            Category c = catMap.get(t.getCategoryId());
            if (c != null) {
                Map<String, Object> cMap = new HashMap<>();
                cMap.put("_id", c.getId());
                cMap.put("name", c.getName());
                cMap.put("icon", c.getIcon());
                cMap.put("color", c.getColor());
                cMap.put("type", c.getType());
                map.put("categoryId", cMap);
            } else {
                map.put("categoryId", t.getCategoryId());
            }
            resultList.add(map);
        }

        Map<String, Object> response = new HashMap<>();
        response.put("transactions", resultList);
        response.put("totalPages", (int) Math.ceil((double) total / limit));
        response.put("currentPage", page);
        response.put("total", total);

        return ResponseEntity.ok(response);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateTransaction(@PathVariable String id, @RequestBody Transaction updates) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        Optional<Transaction> tOpt = transactionRepository.findById(id);
        
        if (tOpt.isEmpty() || !tOpt.get().getUserId().equals(userId)) {
            return ResponseEntity.status(404).body(Map.of("message", "Transaction not found"));
        }
        
        Transaction t = tOpt.get();
        if (updates.getAmount() != 0) t.setAmount(updates.getAmount());
        if (updates.getCategoryId() != null) t.setCategoryId(updates.getCategoryId());
        if (updates.getDate() != null) t.setDate(updates.getDate());
        if (updates.getNote() != null) t.setNote(updates.getNote());
        if (updates.getType() != null) t.setType(updates.getType());
        
        transactionRepository.save(t);
        return ResponseEntity.ok(t);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteTransaction(@PathVariable String id) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        Optional<Transaction> tOpt = transactionRepository.findById(id);
        
        if (tOpt.isEmpty() || !tOpt.get().getUserId().equals(userId)) {
            return ResponseEntity.status(404).body(Map.of("message", "Transaction not found"));
        }
        
        transactionRepository.deleteById(id);
        return ResponseEntity.ok(Map.of("message", "Transaction deleted successfully"));
    }
}
